### PR TITLE
The EUC-TW encoding doesn't have a canonical name

### PR DESCRIPTION
**Pull request by litherum**
#### 996fc9db28be5ace21aed6dfccc9290d3f87d142
<pre>
The EUC-TW encoding doesn&apos;t have a canonical name
<a href="https://bugs.webkit.org/show_bug.cgi?id=246242">https://bugs.webkit.org/show_bug.cgi?id=246242</a>
rdar://100893154

Reviewed by Alex Christensen.

ucnv_getCanonicalName(&quot;EUC-TW&quot;, &quot;IANA&quot;) is returning nullptr, which is causing
us to memcmp(nullptr, ...). This patch reorganizes the file to call ucnv_getName()
in the situations where ucnv_getCanonicalName() is returning nullptr, thereby
elliminating the need for all the special cases.

Test: imported/w3c/web-platform-tests/encoding/unsupported-labels.html

* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerCodecs):
(PAL::TextCodecICU::TextCodecICU):

Canonical link: <a href="https://commits.webkit.org/256438@main">https://commits.webkit.org/256438@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/df96a4ffc93e3a0bb7f3725fb9d52564e5b91240

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95733 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105309 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165632 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99716 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5066 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33746 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88113 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101149 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3726 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82349 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30780 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87489 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73609 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39478 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37173 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4451 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41239 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39596 "Passed tests") | |

### CODE REVIEW

**Comment by aproskuryakov** [comment]
Can this have a regression test?

**Comment by aproskuryakov** [comment]
The code change looks good. One thing that's a little unclear is whether `windows-949` actually gets us `windows-949-2000`, and will in the future. Depending on future ICU decisions seems suboptimal.

**Comment by annevk** [comment]
I'm not sure I entirely understand what this is doing. E.g., windows-949 maps to EUC-KR and as far as I can tell we don't use ICU for that, we use `EncodingTables.cpp`. See also bc3715d824ffde84c98cd0e4870676410ac055a8.

I suspect Alex would be a good reviewer here.

File: Source/WebCore/PAL/pal/text/TextCodecICU.cpp [hunk 1]
```diff
@@ -115,42 +115,15 @@ void TextCodecICU::registerCodecs(TextCodecRegistrar registrar)
...
-            });
-            continue;
-        }
-
         UErrorCode error = U_ZERO_ERROR;
         const char* canonicalConverterName = ucnv_getCanonicalName(name, "IANA", &error);
         ASSERT(U_SUCCESS(error));
+        if (!canonicalConverterName) {
+            auto converter = ICUConverterPtr { ucnv_open(name, &error) };
+            assert(U_SUCCESS(error));
```
Comment (achristensen07) [inline Source/WebCore/PAL/pal/text/TextCodecICU.cpp]: Lowercase assert is a different function that we don't use in WebKit.

File: Source/WebCore/PAL/pal/text/TextCodecICU.cpp [hunk 2]
```diff
@@ -115,42 +115,15 @@ void TextCodecICU::registerCodecs(TextCodecRegistrar registrar)
...
+            canonicalConverterName = ucnv_getName(converter.get(), &error);
+            assert(U_SUCCESS(error));
+        }
         registrar(name, [name, canonicalConverterName] {
```
Comment (achristensen07) [inline Source/WebCore/PAL/pal/text/TextCodecICU.cpp:131]: It probably wouldn't hurt to add another if (!canonicalConverterName) { ASSERT_NOT_REACHED(); continue; } to be on the safe side.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

**Comment by litherum** [comment]
> One thing that's a little unclear is whether `windows-949` actually gets us `windows-949-2000`

It does.

> and will in the future

It won't, but Apple's ICU team views behavior changes here as a progression, rather than a regression.

Also, Chrome is way more liberal than we are:

```
void TextCodecICU::RegisterCodecs(TextCodecRegistrar registrar) {
  // See comment above in registerEncodingNames.
  registrar("ISO-8859-8-I", Create, nullptr);

  const bool is_text_codec_cjk_enabled =
      base::FeatureList::IsEnabled(blink::features::kTextCodecCJKEnabled);
  int32_t num_encodings = ucnv_countAvailable();
  for (int32_t i = 0; i < num_encodings; ++i) {
    const char* name = ucnv_getAvailableName(i);
    UErrorCode error = U_ZERO_ERROR;
    const char* standard_name = ucnv_getStandardName(name, "MIME", &error);
    if (!U_SUCCESS(error) || !standard_name) {
      error = U_ZERO_ERROR;
      standard_name = ucnv_getStandardName(name, "IANA", &error);
      if (!U_SUCCESS(error) || !standard_name)
        continue;
    }
#if defined(USING_SYSTEM_ICU)
    // Explicitly do not support UTF-32. https://crbug.com/417850
    // Bundled ICU does not return these names.
    if (!strcmp(standard_name, "UTF-32") ||
        !strcmp(standard_name, "UTF-32LE") ||
        !strcmp(standard_name, "UTF-32BE")) {
      continue;
    }
#endif
    if (is_text_codec_cjk_enabled && TextCodecCJK::IsSupported(standard_name)) {
      continue;
    }
    registrar(standard_name, Create, nullptr);
  }
}
```

**Review by achristensen07 — APPROVED** [review]

**Pull request closed without merging**